README.md
# NuORDER

[![Gem Version](https://badge.fury.io/rb/nuorder.svg)](http://badge.fury.io/rb/nuorder)
[![Build Status](https://travis-ci.org/springboardretail/nuorder.svg)](https://travis-ci.org/springboardretail/nuorder)

Gem that wraps up the communication with the API services from [NuORDER](http://www.nuorder.com/)

[Faraday](https://github.com/lostisland/faraday) and [Excon](https://github.com/excon/excon) are used for creating persistent connections.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'nuorder'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install nuorder

## Usage

### Configuration

The following options are available when configuring NuORDER:

1. `app_name` Default: `'Springboard Retail'`
2. `api_endpoint` Default: `'http://buyer.nuorder.com'`
3. `oauth_consumer_key`
4. `oauth_consumer_secret`
5. `oauth_token`
6. `oauth_token_secret`
7. `oauth_callback` Default: `'oob'`
    * NuORDER will POST the callback the `oauth_verifier` token.
    * If the callback is set to `'oob'`, the `oauth_verifier` token must be supplied manually by the user.

#### Example

```ruby
Nuorder.configure do |config|
  config.oauth_consumer_key = 'key'
  config.oauth_consumer_secret = 'key'
  config.oauth_token = 'key'
  config.oauth_token_secret = 'key'
end
````

or

```
client = Nuorder::Client.new(oauth_token: 'key', oauth_token_secret: 'key')
```

### Authorization (OAuth 1.0)

NuORDER uses OAuth 1.0 for authorization.

To get the oauth tokens the following methods must be called:

1. `Nuorder.api_initiate`
  * An inital call will be made in order to get the temporary tokens.
2. `Nuorder.get_oauth_token(oauth_verifier)`
  * Call for getting the permanent tokens.
  * `oauth_verifier` must be supplied.

Retrieve tokens:

```ruby
oauth_token = Nuorder.oauth_token
oauth_token_secret = Nuorder.oauth_token_secret
```

### Requests
TODO

## Contributing

1. Fork it ( https://github.com/springboardretail/nuorder/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request

### Running tests

Place your consumer keys and tokens in the `.env` file

```
OAUTH_CONSUMER_KEY = 'nuorder-consumer-key'
OAUTH_CONSUMER_SECRET = 'nuorder-consumer-secret'
OAUTH_TOKEN = 'nuorder-oauth-token'
OAUTH_TOKEN_SECRET = 'nuorder-token-secret'
```

`rake` default task will run all the tests


CHANGELOG.md
## v1.1.0

* Allow multiple clients to be used simultaneously

## v1.0.0

* Initial release


spec/spec_helper.rb
require 'nuorder'
require 'byebug'
require 'vcr'
require 'dotenv'

Dotenv.load

Dir['./spec/support/**/*.rb'].each { |f| require f }


spec/support/vcr.rb
SENSITIVE_DATA = %w(
  OAUTH_CONSUMER_KEY
  OAUTH_CONSUMER_SECRET
  OAUTH_TOKEN
  OAUTH_TOKEN_SECRET
)

VCR.configure do |config|
  config.cassette_library_dir = 'spec/vcr'
  config.hook_into :excon
  config.configure_rspec_metadata!
  
  SENSITIVE_DATA.each do |key|
    config.filter_sensitive_data(key) { ENV[key] }
  end
end


spec/nuorder_spec.rb
require 'spec_helper'

describe Nuorder do
  describe '#new' do
    it 'creates a Nuorder::Client' do
      expect(Nuorder.new).to be_kind_of Nuorder::Client
    end
  end

  describe '.client' do
    it 'creates a Nuorder::Client' do
      expect(Nuorder.client).to be_kind_of Nuorder::Client
    end

    it 'caches the client when the same options are passed' do
      expect(Nuorder.client).to eq(Nuorder.client)
    end

    it 'returns a fresh client when options are not the same' do
      client = Nuorder.client
      Nuorder.oauth_token = '87614b09dd141c22800f96f11737ade5226d7ba8'
      client_two = Nuorder.client
      client_three = Nuorder.client
      expect(client).not_to eq(client_two)
      expect(client_three).to eq(client_two)
    end
  end

  describe '.configure' do
    Nuorder::Configurable.keys.each do |key|
      it "sets the #{key.to_s.gsub('_', ' ')}" do
        Nuorder.configure do |config|
          config.send("#{key}=", key)
        end
        expect(Nuorder.instance_variable_get(:"@#{key}")).to eq(key)
      end
    end
  end
end


lib/nuorder/connector.rb
require 'faraday'
require 'excon'
require 'faraday_middleware-multi_json'

module Nuorder
  class Connector
    attr_reader :oauth

    def initialize(app_name, base_uri, config)
      @app_name = app_name
      @base_uri = base_uri

      @oauth = Oauth.new(app_name, base_uri, config)
    end

    def api_initiate
      addons = { 'application_name' => @app_name, 'oauth_callback' => @oauth.callback }
      headers = @oauth.headers('GET', '/api/initiate', addons)
      response = connection.get '/api/initiate', {}, headers

      oauth.token = response.body['oauth_token']
      oauth.token_secret = response.body['oauth_token_secret']

      response
    end

    def get_oauth_token(oauth_verifier)
      fail 'No oauth_verifier' unless oauth_verifier

      headers = oauth.headers('GET', '/api/token', { 'oauth_verifier' => oauth_verifier })
      response = connection.get '/api/token', {}, headers

      oauth.token = response.body['oauth_token']
      oauth.token_secret = response.body['oauth_token_secret']

      response
    end

    def get(url, params: nil)
      headers = oauth.headers('GET', url)
      response = connection.get url, params, headers
      validate_response(response)
    end

    private

    def connection
      @connection ||= build_connection
    end

    def build_connection
      Faraday.new(url: @base_uri) do |builder|
        builder.response :multi_json
        builder.adapter :excon, persistent: true
      end
    end

    def validate_response(response)
      return response if [200, 201].include?(response.status)
      fail ApiError, "NuOrder API Error #{response.body['code']}. #{response.body['message']}"
    end
  end
end


lib/nuorder/client/oauth.rb
require 'securerandom'

module Nuorder
  class Client
    module Oauth
      VERSION = '1.0'
      SIGNATURE_METHOD = 'HMAC-SHA1'

      def oauth_headers(method, url, addons = nil)
        time = Time.now.to_i
        nonce = SecureRandom.hex(8)
        signature = build_signature(method, url, time, nonce, addons)
        oauth_header = build_oauth(time, nonce, signature, addons)
        {
            'Authorization' => "oAuth #{oauth_header}",
            'Accept' => 'application/json',
            'Content-Type' => 'application/json'
        }
      end

      private

      def build_signature(method, url, time, nonce, addons)
        base = "#{method}#{@api_endpoint}#{url}?"
        base << "oauth_consumer_key=#{@oauth_consumer_key}"
        base << "&oauth_token=#{@oauth_token}"
        base << "&oauth_timestamp=#{time}"
        base << "&oauth_nonce=#{nonce}"
        base << "&oauth_version=#{VERSION}&oauth_signature_method=#{SIGNATURE_METHOD}"
        base << "&oauth_verifier=#{addons['oauth_verifier']}" if addons && addons.include?('oauth_verifier')
        base << "&oauth_callback=#{addons['oauth_callback']}" if addons && addons.include?('oauth_callback')
        key = [@oauth_consumer_secret, @oauth_token_secret].join('&')
        digest = OpenSSL::Digest.new('sha1')
        OpenSSL::HMAC.hexdigest(digest, key, base)
      end

      def build_oauth(time, nonce, signature, addons)
        oauth = {
            'oauth_consumer_key' => @oauth_consumer_key,
            'oauth_timestamp' => time,
            'oauth_nonce' => nonce,
            'oauth_version' => VERSION,
            'oauth_signature_method' => SIGNATURE_METHOD,
            'oauth_token' => @oauth_token,
            'oauth_signature' => signature
        }
        oauth.merge!(addons) unless addons.nil?
        oauth.map { |k, v| "#{k}=#{v}" }.join(',')
      end
    end
  end
end


lib/nuorder/configurable.rb
module Nuorder
  module Configurable
    class << self
      def keys
        @keys ||= %i(
          app_name
          api_endpoint
          oauth_callback
          oauth_consumer_key
          oauth_consumer_secret
          oauth_token
          oauth_token_secret
        )
      end
    end

    attr_accessor *Nuorder::Configurable.keys

    # Set configuration options using a block
    def configure
      yield(self) if block_given?
    end

    # Reset configuration options to default values
    def reset!
      Nuorder::Configurable.keys.each do |key|
        instance_variable_set(:"@#{key}", Nuorder::Default.options[key])
      end
      self
    end
    alias_method :setup, :reset!

    def options
      Hash[Nuorder::Configurable.keys.map{|key| [key, instance_variable_get(:"@#{key}")]}]
    end
  end
end


lib/nuorder/client.rb
require 'faraday'
require 'excon'
require 'faraday_middleware-multi_json'
require 'nuorder/configurable'
require 'nuorder/client/oauth'

module Nuorder
  class Client

    include Nuorder::Configurable
    include Nuorder::Client::Oauth

    def initialize(options = {})
      options = Nuorder::Default.options.merge(options)
      set_instance_variables(options)
    end

    def same_options?(opts)
      opts.hash == options.hash
    end

    def api_initiate
      addons = { 'application_name' => @app_name, 'oauth_callback' => @oauth_callback }
      headers = oauth_headers('GET', '/api/initiate', addons)
      response = connection.get '/api/initiate', {}, headers

      @oauth_token = response.body['oauth_token']
      @oauth_token_secret = response.body['oauth_token_secret']

      response
    end

    def get_oauth_token(oauth_verifier)
      fail 'No oauth_verifier' unless oauth_verifier

      headers = oauth_headers('GET', '/api/token', { 'oauth_verifier' => oauth_verifier })
      response = connection.get '/api/token', {}, headers

      @oauth_token = response.body['oauth_token']
      @oauth_token_secret = response.body['oauth_token_secret']

      response
    end

    def get(url, params: nil)
      headers = oauth_headers('GET', url)
      response = connection.get url, params, headers
      validate_response(response)
    end

    def orders(status:)
      get("/api/orders/#{status}/detail").body
    end

    def product(id:)
      get("/api/product/#{id}").body
    end

    def orders_with_product_details(status:)
      orders = orders(status: status)
      orders.map! do |order|
        add_product_details!(line_items: order['line_items'])
        order
      end
    end

    private

    # Set instance variables passed in options
    def set_instance_variables(options)
      Nuorder::Configurable.keys.each do |key|
        instance_variable_set(:"@#{key}", options[key])
      end
    end

    def connection
      @connection ||= build_connection
    end

    def build_connection
      Faraday.new(url: @api_endpoint) do |builder|
        builder.response :multi_json
        builder.adapter :excon, persistent: true
      end
    end

    def validate_response(response)
      return response if [200, 201].include?(response.status)
      fail ApiError, "NuOrder API Error #{response.body['code']}. #{response.body['message']}"
    end

    def add_product_details!(line_items:)
      line_items.map! do |line_item|
        product_id = line_item['product']['_id']
        product_details = product(id: product_id)
        line_item['product'].merge!(product_details)
        line_item
      end
    end
  end
end
